### PR TITLE
feat(bitswap): add `validate_peer` callback to bitswap request manager

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As a follow-up of https://github.com/ChainSafe/forest/pull/3713
When a forest node is sending a bitswap request, there's no point in sending it to peers that are out of sync or paused for sync. This change should mitigate the `bitswap` timeout when a lot of peers are connected to a forest node.

Changes introduced in this pull request:

- Add `validate_peer` callback to bitswap request manager to allow custom peer filtering on the caller side to boost performance
- Filter out peers whose head epochs are older than the epoch of the requested bitswap message.
- Refactor `tokio::sync::RwLock` in `PeerManager` into `parking_lot::RwLock`

Manually tested on mainnet for a while, got 100% bitswap success rate
```
# TYPE bitswap_message_count counter
bitswap_message_count{type="get_block_success"} 1798
bitswap_message_count{type="inbound_request_have"} 645
bitswap_message_count{type="inbound_response_block"} 178581
bitswap_message_count{type="inbound_response_block_already_exists_in_db"} 11
bitswap_message_count{type="inbound_response_block_not_requested"} 176818
bitswap_message_count{type="inbound_response_block_update_db"} 1731
bitswap_message_count{type="inbound_response_have_yes"} 80512
bitswap_message_count{type="inbound_stream_count"} 157267
bitswap_message_count{type="outbound_request_block"} 2334
bitswap_message_count{type="outbound_request_cancel"} 744345
bitswap_message_count{type="outbound_request_have"} 537571
bitswap_message_count{type="outbound_response_have"} 226
bitswap_message_count{type="outbound_stream_count"} 617836
```

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/libp2p/service.rs [hunk 1]
```diff
@@ -425,8 +428,25 @@ async fn handle_network_message(
...
             response_channel,
+            epoch,
         } => {
-            bitswap_request_manager.get_block(store, cid, BITSWAP_TIMEOUT, Some(response_channel));
+            bitswap_request_manager.get_block(
+                store,
+                cid,
+                BITSWAP_TIMEOUT,
+                Some(response_channel),
+                if let Some(epoch) = epoch {
```
Comment (LesnyRumcajs) [inline src/libp2p/service.rs:438]: Let's move it to a variable. This is too complex to embed as a param.

File: src/libp2p_bitswap/request_manager.rs [hunk 2]
```diff
@@ -162,6 +163,7 @@ impl BitswapRequestManager {
         store: Arc<impl BitswapStoreReadWrite>,
         cid: Cid,
         deadline: Instant,
+        validate_peer: Option<Arc<dyn Fn(PeerId) -> bool + Send + Sync>>,
```
Comment (LesnyRumcajs) [inline src/libp2p_bitswap/request_manager.rs:166]: For readability, I'd make it a named type, e.g., `PeerValidateCallback`. What do you think?

Comment (hanabi1224) [reply src/libp2p_bitswap/request_manager.rs:166]: Fixed

File: src/libp2p/service.rs [hunk 3]
```diff
@@ -425,8 +429,27 @@ async fn handle_network_message(
...
+                None
+            };
+
+            bitswap_request_manager.get_block(
+                store,
+                cid,
+                BITSWAP_TIMEOUT,
+                Some(response_channel),
+                validate_peer,
+            );
```
Comment (LesnyRumcajs) [inline src/libp2p/service.rs:452]:
```suggestion
            let peer_validator: Option<Arc<ValidatePeerCallback>> = if let Some(epoch) = epoch {
                let peer_manager = Arc::clone(peer_manager);
                Some(Arc::new(move |peer| {
                    peer_manager
                        .get_peer_head_epoch(&peer)
                        .map(|peer_head_epoch| peer_head_epoch >= epoch)
                        .unwrap_or_default()
                }))
            } else {
                None
            };

            bitswap_request_manager.get_block(
                store,
                cid,
                BITSWAP_TIMEOUT,
                Some(response_channel),
                peer_validator,
            );
```

**Review by LesnyRumcajs — APPROVED** [review]

**Review by lemmih — APPROVED** [review]

**Pull request merged**